test-case/check-audio-equalizer.sh

```diff
@@ -0,0 +1,146 @@
+#!/bin/bash
+
+##
+## Case Name: check-audio-equalizer.sh
+## Preconditions:
+##    SOF Topology should have EQ component(s) in any pipeline
+## Description:
+##    check IIR/FIR config working
+## Case step:
+##    1. Check if the topology has a EQ component in it
+##    2. Test with IIR config list
+##    3. Test with FIR config list
+## Expect result:
+##    sof-ctl and aplay should return 0
+##
+
+# source from the relative path of current folder
+source $(dirname ${BASH_SOURCE[0]})/../case-lib/lib.sh
+
+OPT_OPT_lst['t']='tplg'     OPT_DESC_lst['t']='tplg file, default value is env TPLG: $TPLG'
+OPT_PARM_lst['t']=1         OPT_VALUE_lst['t']="$TPLG"
+
+OPT_OPT_lst['d']='duration' OPT_DESC_lst['d']='aplay duration in second'
+OPT_PARM_lst['d']=1         OPT_VALUE_lst['d']=5
+
+OPT_OPT_lst['l']='loop'     OPT_DESC_lst['l']='loop count'
+OPT_PARM_lst['l']=1         OPT_VALUE_lst['l']=1
+
+OPT_OPT_lst['s']='sof-logger'   OPT_DESC_lst['s']="Open sof-logger trace the data will store at $LOG_ROOT"
+OPT_PARM_lst['s']=0             OPT_VALUE_lst['s']=1
+
+func_opt_parse_option $*
+tplg=${OPT_VALUE_lst['t']}
+duration=${OPT_VALUE_lst['d']}
+loop_cnt=${OPT_VALUE_lst['l']}
+
+func_pipeline_export $tplg "type:playback"
+sofcard=${SOFCARD:-0}
+
+# Test equalizer
+func_test_eq()
+{
+    local id=$1
+    local conf=$2
+    dlogc "sof-ctl -Dhw:$sofcard -n $id -s $conf"
+    [[ ! -f $conf ]] && dloge" $conf not exist";exit 1
+
+    sof-ctl -Dhw:0 -n $id -s $conf
+    if [ $? -ne 0 ]; then
+        dloge "Equalizer setting failure with $conf"
+        return 1
+    fi
+    # test with aplay
+    dlogc "aplay -D$dev -f $fmt -c $channel /dev/zero -d $duration"
+    aplay -D$dev -f $fmt -c $channel /dev/zero -d $duration
+    if [ $? -ne 0 ]; then
+        dloge "Equalizer test failure with $conf"
+        return 1
+    fi
+    sleep 1
+}
+
+
+eqConfRoot="./eqctl"
+# initialized with negative value. If none of test is performed the test result will be failed.
+result=1
+failed_cnt=0
+
+dlogi "0. Get amixer control id for IIR and FIR"
+IIRid=`amixer -D hw:0 controls | grep EQIIR| head -1| sed 's/numid=\([0-9]*\),.*/\1/'`
+if [ -z $IIRid ]; then
+        dloge "can't find IIR filter"
+        exit 1
+fi
+
+FIRid=`amixer -D hw:0 controls | grep EQFIR| head -1| sed 's/numid=\([0-9]*\),.*/\1/'`
+if [ -z $FIRid ]; then
+        dloge "can't find FIR filter"
+        exit 1
+fi
+
+
+for idx in $(seq 0 $(expr $PIPELINE_COUNT - 1))
```

Comment:
Please check this PR for this kind of idea. If you think it's useful to pass the EQ response from top level topology with pipeline.m4 go ahead and copy it https://github.com/thesofproject/sof/pull/2434 . It doesn't make sense to make a new eq pipeline macro for every new needed response. All see how I make the EQ control bytes and tokens (also volume tokens) unique for pipeline instance. Without such it's not possible initalize in topology equalizers to different response because the topology.conf file has "flat" hierarchy.

I would like to do this kind of cleanup anyway but I'm pretty busy with bugs so as my task it would take some time.

Reply:
Sorry, I thought this was @fredoh9 's EQ topology proposal. My comment belongs there and to EQ test script.

Reply:
@xiulipan i will work with @aiChaoSONG for this new feature.

Reply:
Could you work with @Bin-QA @aiChaoSONG to directly get tplg with EQ here. And skip if the the PCM list is empty.